cpp/src/rolling/nth_element.cuh

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/aggregation.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/utilities/bit.hpp>
+
+#include <limits>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf::detail::rolling {
+
+auto constexpr NULL_INDEX = std::numeric_limits<size_type>::min();  // For nullifying with gather.
+
+template <null_policy null_handling, typename PrecedingIter, typename FollowingIter>
+std::unique_ptr<column> nth_element(size_type n,
+                                    column_view const& input,
+                                    PrecedingIter preceding,
+                                    FollowingIter following,
+                                    size_type min_periods,
+                                    rmm::cuda_stream_view stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  auto const gather_iter = cudf::detail::make_counting_transform_iterator(
```

Comment:
The lambda below is sooo huge. Can we break this into multiple steps? I suspect that register usage for such giant kernel will be significant, which may lead to poor performance.

Reply:
Your suggestion is to break the lambda up into functions?

Reply:
I meant breaking the `make_counting_transform_iterator` into several smaller `transform` steps, but I'm not sure if that is more efficient than just having one big transform step like this...

Reply:
I would propose making this into a functor instead of an inline lambda to make it easier to digest.

Reply:
I'm glad you both insisted on breaking this up. The result is a little more verbose (capture vs constructor), but it's more readable.